### PR TITLE
Orca: Make optimizer_creator optional and discuss optimizers when fit and eval in pytorch estimator

### PR DESCRIPTION
**Pull request by zpeng1898**
## Description

When in predict mode, there should be no optimizer_creator. So we should require optimizer_creator when fiting or validating instead of init estimator.

[issue165](https://github.com/analytics-zoo/orca/issues/165#issuecomment-1439630520)

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 1]
```diff
@@ -449,6 +453,9 @@ def validate(self, data_creator, batch_size=32, num_steps=None, profile=False,
         if not self.criterion:
             invalidInputError(False,
                               "You must provide a loss for train and evaluate.")
+        if not self.optimizers:
+            invalidInputError(False,
+                              "You must provide the optimizer for train and evaluate.")
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:458]: Does evaluation need an optimizer?

**Review by leonardozcm — APPROVED** [review]
Others LGTM

**Comment by hkvision** [comment]
Also add a simple unit test to cover this change?

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 2]
```diff
@@ -278,10 +278,18 @@ def train_epoch(self,
...
+            invalidInputError(False,
+                              "You must provide a model for train and evaluate.")
+
         if not self.criterion:
             invalidInputError(False,
                               "You must provide a loss for train and evaluate.")
 
+        if not self.optimizers:
+            invalidInputError(False,
+                              "You must provide the optimizer for train and evaluate.")
```
Comment (sgwhat) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:291]: Maybe remove `and evaluate`? 😄

**Review by sgwhat — APPROVED** [review]

**Comment by hkvision** [comment]
seems one unit test failed, take a look @zpeng1898 
```
FAILED test/bigdl/orca/learn/ray/pytorch/test_estimator_ray_backend.py::TestPyTorchEstimator::test_optional_model_creator - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/entire_model.pth'
============= 1 failed, 23 passed, 4 warnings in 209.36s (0:03:29) =============
```

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 3]
```diff
@@ -581,11 +593,15 @@ def forward_batch(self, batch, callbacks=None):
...
     def predict(self, partition, batch_size=32, profile=False, callbacks=None):
-        """Evaluates the model on the validation data set."""
+        """Predict the model."""
         config = copy.copy(self.config)
         self._toggle_profiling(profile=profile)
         callbacks = callbacks or []
 
+        if not self.models:
+            invalidInputError(False,
+                              "You must provide a model for predicting.")
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:603]: predicting -> predict

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 4]
```diff
@@ -278,10 +278,18 @@ def train_epoch(self,
             data_loader.sampler.set_epoch(self.epochs)
         self.logger.debug("Begin Training Step {}".format(self.epochs + 1))
 
+        if not self.models:
+            invalidInputError(False,
+                              "You must provide a model for train and evaluate.")
+
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:284]: since model creator is not optional, we don't need this?

Comment (hkvision) [reply python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:284]: @sgwhat Align whether model creator is optional for spark/ray backends.

**Pull request merged**